File: src/main/java/com/zoarial/iot/network/IoTPacketSection.java
package com.zoarial.iot.network;

import com.zoarial.iot.model.IoTBasicType;

import java.nio.ByteBuffer;
import java.util.UUID;

public class IoTPacketSection {
    final private IoTBasicType type;
    final private String str;
    final private long num;
    final private long uuid_low;

    public IoTPacketSection(String str) {
        if (str.equals("ZIoT")) {
            type = IoTBasicType.HEADER;
        } else {
            type = IoTBasicType.STRING;
        }
        this.str = str;
        num = 0;
        uuid_low = 0;
    }

    public IoTPacketSection(byte b) {
        type = IoTBasicType.BYTE;
        num = (int)b & 0xFF;
        uuid_low = 0;
        str = "";
    }

    public IoTPacketSection(int i) {
        type = IoTBasicType.INTEGER;
        num = i;
        uuid_low = 0;
        str = "";
    }

    public IoTPacketSection(long i) {
        type = IoTBasicType.LONG;
        num = i;
        uuid_low = 0;
        str = "";
    }

    public IoTPacketSection(UUID uuid) {
        type = IoTBasicType.UUID;
        num = uuid.getMostSignificantBits();
        uuid_low = uuid.getLeastSignificantBits();
        str = "";
    }

    public IoTPacketSection(boolean b) {
        type = IoTBasicType.BOOLEAN;
        num = 0;
        uuid_low = 0;
        str = b ? "true" : "false";
    }

    public byte[] getByteList() {
        return switch (type) {
            case BYTE -> new byte[]{(byte) num};
            case LONG -> ByteBuffer.allocate(8).putLong(num).array();
            case UUID -> ByteBuffer.allocate(16).putLong(num).putLong(uuid_low).array();
            //TODO: Decided if i want to add JSONObject to hold the JSON data
            case BOOLEAN, STRING, JSON -> {
                byte[] strArr = str.getBytes();
                int len = strArr.length;
                byte[] byteArr = new byte[strArr.length + 1];
                System.arraycopy(strArr, 0, byteArr, 0, len);
                byteArr[len] = 0; // Set null byte
                yield byteArr;
            }
            case INTEGER -> ByteBuffer.allocate(4).putInt((int) num).array();
            case HEADER -> str.getBytes();
        };
    }

    public IoTBasicType getType() {
        return type;
    }

}

File: src/main/java/com/zoarial/iot/network/IoTPacketSection.kt
package com.zoarial.iot.network

import com.zoarial.iot.model.IoTBasicType
import java.nio.ByteBuffer
import java.util.*

class IoTPacketSection {
    val type: IoTBasicType
    private val str: String?
    private val num: Long
    private val uuid_low: Long

    constructor(str: String?) {
        type = if (str == "ZIoT") {
            IoTBasicType.HEADER
        } else {
            IoTBasicType.STRING
        }
        this.str = str
        num = 0
        uuid_low = 0
    }

    constructor(b: Byte) {
        type = IoTBasicType.BYTE
        num = (b.toInt() and 0xFF).toLong()
        uuid_low = 0
        str = ""
    }

    constructor(i: Int) {
        type = IoTBasicType.INTEGER
        num = i.toLong()
        uuid_low = 0
        str = ""
    }

    constructor(i: Long) {
        type = IoTBasicType.LONG
        num = i
        uuid_low = 0
        str = ""
    }

    constructor(uuid: UUID) {
        type = IoTBasicType.UUID
        num = uuid.mostSignificantBits
        uuid_low = uuid.leastSignificantBits
        str = ""
    }

    constructor(b: Boolean) {
        type = IoTBasicType.BOOLEAN
        num = (if (b) 1 else 0).toLong()
        uuid_low = 0
        str = ""
    }

    // Set null byte
    val byteList: ByteArray
        get() = when (type) {
            IoTBasicType.BYTE -> byteArrayOf(num.toByte())
            IoTBasicType.LONG -> ByteBuffer.allocate(8).putLong(num).array()
            IoTBasicType.UUID -> ByteBuffer.allocate(16).putLong(num).putLong(uuid_low).array()
            IoTBasicType.BOOLEAN -> ByteBuffer.allocate(1).put(num.toByte()).array()
            IoTBasicType.STRING, IoTBasicType.JSON -> {
                val strArr = str!!.toByteArray()
                val len = strArr.size
                val byteArr = ByteArray(strArr.size + 1)
                System.arraycopy(strArr, 0, byteArr, 0, len)
                byteArr[len] = 0 // Set null byte
                byteArr
            }

            IoTBasicType.INTEGER -> ByteBuffer.allocate(4).putInt(num.toInt()).array()
            IoTBasicType.HEADER -> str!!.toByteArray()
        }
}